crates/dht/src/hash.rs

```diff
@@ -0,0 +1,443 @@
+    async fn create_with_default_partition_factor() {
+        enable_tracing();
+
+        let store = Arc::new(Kitsune2MemoryOpStore::default());
+        let ph =
+            PartitionedHashes::try_from_store(20, 14, Timestamp::now(), store)
+                .await
+                .unwrap();
+        assert_eq!(1 << 20, ph.size);
+        assert_eq!((0, true), (1u32 << 20).overflowing_mul(ph.size));
+    }
+
+    #[tokio::test]
+    async fn covers_full_arc() {
+        enable_tracing();
+
+        // Check that the behaviour is consistent for a reasonable range of expected values
+        for hash_factor in 20u8..31 {
+            let store = Arc::new(Kitsune2MemoryOpStore::default());
+            let ph = PartitionedHashes::try_from_store(
+                hash_factor,
+                14,
+                Timestamp::now(),
+                store,
+            )
+            .await
+            .unwrap();
+
+            let mut start: u32 = 0;
+            for i in 0..ph.partitioned_hashes.len() {
+                let end = start.overflowing_add(ph.size).0;
+                assert_eq!(start, ph.partitioned_hashes[i]._arc.0);
+                assert_eq!(end, ph.partitioned_hashes[i]._arc.1);
+                start = end;
+            }
+
+            assert_eq!(0, start, "While checking factor {}", hash_factor);
+        }
+    }
+
+    #[tokio::test]
+    async fn inform_ops_stored_in_full_slices() {
+        enable_tracing();
+
+        let store = Arc::new(Kitsune2MemoryOpStore::default());
+        let mut ph = PartitionedHashes::try_from_store(
+            20,
+            14,
+            Timestamp::now(),
+            store.clone(),
+        )
+        .await
+        .unwrap();
+
+        let op_id_bytes_1 = bytes::Bytes::from_static(&[7, 0, 0, 0]);
+        let op_id_bytes_2 = bytes::Bytes::from(ph.size.to_le_bytes().to_vec());
+        ph.inform_ops_stored(
+            store.clone(),
+            vec![
+                StoredOp {
+                    op_id: OpId::from(op_id_bytes_1.clone()),
+                    timestamp: UNIX_TIMESTAMP,
+                },
+                StoredOp {
+                    op_id: OpId::from(op_id_bytes_2.clone()),
+                    timestamp: UNIX_TIMESTAMP
+                        + ph.partitioned_hashes[0]
+                            .partitioned_time
+                            .full_slice_duration(),
+                },
+            ],
+        )
+        .await
+        .unwrap();
+
+        let count = store.slice_hash_count((0, ph.size)).await.unwrap();
+        assert_eq!(1, count);
+
+        let hash = store.retrieve_slice_hash((0, ph.size), 0).await.unwrap();
+        assert!(hash.is_some());
+        assert_eq!(op_id_bytes_1, hash.unwrap());
+
+        let count = store
+            .slice_hash_count((ph.size, 2 * ph.size))
+            .await
+            .unwrap();
+        // Note that this is because we've stored at id 1, not that two hashes ended up in this
```

Comment:
Yeah, this naming is confusing. Perhaps it should be `slice_hash_id` instead?

Reply:
It is a count, it's the `highest_stored_slice_id + 1`. I believe the naming is correct here and it's the value that's potentially confusing. That's just because we've stored a combined hash at slice_id 1 and that means we consider ourselves to have 2 total slices.

I can try to rewrite the comment to make that clearer?

Reply:
Is it really a count? There can be gaps in the vector of stored slice ids for the empty slices, so could it be something like `[_, _, _, _, 4, _, 6]`. The count would be `2` for me. The highest slice id is `6`.

Reply:
I say we keep it and a better term will emerge with usage.